File: src/relay/backend/compile_engine.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*!
 * \file relay/backend/compile_engine.cc
 * \brief Internal compilation engine.
 */
#include "compile_engine.h"

#include <tvm/driver/driver_api.h>
#include <tvm/ir/type_functor.h>
#include <tvm/relay/analysis.h>
#include <tvm/relay/attrs/device_copy.h>
#include <tvm/relay/expr.h>
#include <tvm/relay/expr_functor.h>
#include <tvm/relay/op.h>
#include <tvm/relay/op_attr_types.h>
#include <tvm/runtime/registry.h>
#include <tvm/te/operation.h>
#include <tvm/te/schedule.h>
#include <tvm/te/schedule_pass.h>
#include <tvm/topi/tags.h>

#include <functional>
#include <limits>
#include <mutex>
#include <unordered_map>
#include <utility>
#include <vector>

#include "../../runtime/meta_data.h"
#include "../transforms/pass_utils.h"
#include "te_compiler_cache.h"
#include "utils.h"

namespace tvm {
namespace relay {

TVM_REGISTER_OBJECT_TYPE(CompileEngineNode);

class CompileEngineImpl : public CompileEngineNode {
 public:
  // Lower the function.
  CachedFunc Lower(const CCacheKey& key, std::function<String(String)> mangle_fn) {
    return LowerInternal(key, mangle_fn)->cached_func;
  }

  CachedFunc Lower(const CCacheKey& key, const String mod_name) {
    auto mangle_fn = [mod_name](String name) { return runtime::get_name_mangled(mod_name, name); };

    return Lower(key, mangle_fn);
  }

  // For now, build one module per function.
  PackedFunc JIT(const CCacheKey& key) final {
    auto mangle_fn = [](String name) { return name; };
    CCacheValue value = LowerInternal(key, mangle_fn);
    if (value->packed_func != nullptr) return value->packed_func;
    auto m = build(value->cached_func->funcs, key->target, Target(nullptr));
    value->packed_func = m.GetFunction(value->cached_func->prim_fn_var->name_hint);
    return value->packed_func;
  }

  CachedFunc LowerShapeFunc(const CCacheKey& key) final {
    return LowerShapeFuncInternal(key)->cached_func;
  }

  Array<tvm::runtime::Module> LowerExternalFunctions() {
    Array<tvm::runtime::Module> ret;
    std::unordered_map<std::string, std::string> cached_symbol;
    std::vector<CCacheKey> cached_ext_funcs;
    for (const auto& it : cache_) {
      auto src_func = it.first->source_func;
      ICHECK(src_func.defined());

      if (src_func->GetAttr<String>(attr::kCompiler).defined()) {
        auto code_gen = src_func->GetAttr<String>(attr::kCompiler);
        ICHECK(code_gen.defined()) << "No external codegen is set";
        std::string code_gen_name = code_gen.value();
        cached_ext_funcs.push_back(it.first);

        auto symbol_name = src_func->GetAttr<String>(tvm::attr::kGlobalSymbol);
        ICHECK(symbol_name.defined()) << "No external symbol is set for:\n"
                                      << AsText(src_func, false) << "\n"
                                      << "Functions with external codegen must have the "
                                      << tvm::attr::kGlobalSymbol << " attr set.";

        std::string sn = symbol_name.value();
        if (!cached_symbol.count(sn)) {
          cached_symbol[sn] = code_gen_name;
        } else {
          ICHECK_NE(cached_symbol[sn], code_gen_name)
              << "Found duplicated symbol: " << sn << " for: " << code_gen_name;
        }

        std::string ext_name = "relay.ext." + code_gen_name;
        auto pf = tvm::runtime::Registry::Get(ext_name);
        ICHECK(pf) << "Failed to find the codegen tool for " << ext_name << "\n";
        // No need to keep compiler attribute at this point, functions have been
        // extracted for specific codegen.
        src_func = WithAttr(std::move(src_func), attr::kCompiler, NullValue<ObjectRef>());
        runtime::Module ext_mod = (*pf)(src_func);

        // todo(@zhiics, @jroesch): Should this be a user visible error?
        ICHECK(ext_mod.defined()) << "No external library was generated for " << ext_name
                                  << "even though it was requested"
                                     "by the annotated function "
                                  << PrettyPrint(src_func);

        ret.push_back(ext_mod);
      }
    }

    // No need to cache external functions as we collected them all to create
    // external runtime modules.
    for (const auto& it : cached_ext_funcs) {
      cache_.erase(it);
    }
    return ret;
  }

  void Clear() final { cache_.clear(); }

  // List all items in the cache.
  Array<ObjectRef> ListItems() {
    std::lock_guard<std::mutex> lock(mutex_);
    Array<ObjectRef> items;
    for (auto& kv : cache_) {
      items.push_back(kv.first);
      items.push_back(kv.second);
    }
    return items;
  }

  // List all items in the shape_func_cache.
  Array<ObjectRef> ListShapeFuncItems() {
    std::lock_guard<std::mutex> lock(mutex_);
    Array<ObjectRef> items;
    for (auto& kv : shape_func_cache_) {
      items.push_back(kv.first);
      items.push_back(kv.second);
    }
    return items;
  }

  /*!
   * \brief Get the cache key of the function that is being lowered currently
   * \return the cache key
   */
  CCacheKey GetCurrentCCacheKey() { return cur_ccache_key_; }

 private:
  // implement lowered func
  CCacheValue LowerInternal(const CCacheKey& key, std::function<String(String)> mangle_fn) {
    std::lock_guard<std::mutex> lock(mutex_);
    CCacheValue value;
    auto it = cache_.find(key);
    if (it != cache_.end()) {
      it->second->use_count += 1;
      if (it->second->cached_func.defined()) return it->second;
      value = it->second;
    } else {
      value = CCacheValue(make_object<CCacheValueNode>());
      value->use_count = 0;
      if (!backend::IsCompileEngineCacheDisabled()) {
        cache_[key] = value;
      }
    }
    cur_ccache_key_ = key;

    // No need to lower external functions for now. We will invoke the external
    // codegen tool once and lower all functions together.
    if (key->source_func->GetAttr<String>(attr::kCompiler).defined()) {
      auto ir_module = IRModule();
      const auto name_node = key->source_func->GetAttr<String>(tvm::attr::kGlobalSymbol);
      ICHECK(name_node.defined()) << "External function has not been attached a name yet.";
      auto func_name = std::string(name_node.value());
      auto target = Target("ext_dev");
      auto global_var = GlobalVar(func_name);
      global_var->checked_type_ = key->source_func->checked_type();
      ir_module->Add(global_var, key->source_func);
      value->cached_func = CachedFunc(target, global_var, {}, {}, te::Schedule(), {}, ir_module);
      return value;
    }

    // Enforce use the target.
    With<Target> target_scope(key->target);

    ICHECK(!value->cached_func.defined());
    auto cfunc = PrimFuncFor(key->source_func, key->target, [&](std::string name) {
      return GetUniqueName(mangle_fn(name), &name_map_);
    });

    // Skip lowering for device copy node.
    const Expr body = (key->source_func)->body;
    if (const CallNode* call_node = body.as<CallNode>()) {
      if (call_node->attrs.as<DeviceCopyAttrs>()) {
        value->cached_func = cfunc;
        return value;
      }
    }

    // NOTE: array will copy on write.
    Array<te::Tensor> all_args = Array<te::Tensor>(cfunc->inputs);
    for (te::Tensor arg : cfunc->outputs) {
      all_args.push_back(arg);
    }
    // lower the function
    std::unordered_map<te::Tensor, tir::Buffer> binds;
    auto func_name = cfunc->prim_fn_var->name_hint;
    cfunc->funcs->Update(tvm::LowerSchedule(cfunc->schedule, all_args, func_name, binds));
    value->cached_func = cfunc;

    return value;
  }

  // implement lowered shape func
  CCacheValue LowerShapeFuncInternal(const CCacheKey& key) {
    std::lock_guard<std::mutex> lock(mutex_);
    CCacheValue value;
    auto it = shape_func_cache_.find(key);
    if (it != shape_func_cache_.end()) {
      it->second->use_count += 1;
      if (it->second->cached_func.defined()) return it->second;
      value = it->second;
    } else {
      value = CCacheValue(make_object<CCacheValueNode>());
      value->use_count = 0;
      shape_func_cache_[key] = value;
    }
    // Enforce use the target.
    With<Target> target_scope(key->target);

    ICHECK(!value->cached_func.defined());
    using tvm::transform::PassContext;
    With<PassContext> fresh_pass_ctx_scope(PassContext::Create());

    auto cached_func = ShapeFuncFor(key->source_func, key->target, [&](std::string name) {
      return GetUniqueName(name, &name_map_);
    });

    value->cached_func = cached_func;
    return value;
  }

  /*! \brief compiler cache lock*/
  std::mutex mutex_;
  /*! \brief internal name map to get an unique name */
  std::unordered_map<std::string, int> name_map_;
  /*! \brief internal compiler cache */
  std::unordered_map<CCacheKey, CCacheValue> cache_;
  /*! \brief internal compiler cache for shape funcs */
  std::unordered_map<CCacheKey, CCacheValue> shape_func_cache_;
  /*! \brief the cache key of the function that is being lowered currently*/
  CCacheKey cur_ccache_key_;
};

/*! \brief The global compile engine */
CompileEngine& CompileEngine::Global() {
  // intentionally allocate raw pointer to avoid
  // free during destructuion.
  static CompileEngine* inst = new CompileEngine(make_object<CompileEngineImpl>());
  return *inst;
}

TVM_REGISTER_PASS_CONFIG_OPTION("relay.backend.use_auto_scheduler", Bool);
TVM_REGISTER_PASS_CONFIG_OPTION("relay.backend.disable_compile_engine_cache", Bool);

// Make LoweredOutput
TVM_REGISTER_GLOBAL("relay.backend._make_LoweredOutput")
    .set_body_typed([](tvm::Array<te::Tensor> outputs, OpImplementation impl) {
      return LoweredOutput(outputs, impl);
    });

TVM_REGISTER_GLOBAL("relay.backend._make_CCacheKey")
    .set_body_typed([](Function source_func, Target target) {
      return CCacheKey(source_func, target);
    });

TVM_REGISTER_GLOBAL("relay.backend._CompileEngineGlobal").set_body_typed([]() {
  return CompileEngine::Global();
});

TVM_REGISTER_GLOBAL("relay.backend._CompileEngineClear").set_body_typed([](CompileEngine self) {
  self->Clear();
});

TVM_REGISTER_GLOBAL("relay.backend._CompileEngineLower")
    .set_body_typed([](CompileEngine self, CCacheKey key, const String mod_name) {
      return self->Lower(key, mod_name);
    });

TVM_REGISTER_GLOBAL("relay.backend._CompileEngineLowerShapeFunc")
    .set_body_typed([](CompileEngine self, CCacheKey key) { return self->LowerShapeFunc(key); });

TVM_REGISTER_GLOBAL("relay.backend._CompileLowerExternalFunctions")
    .set_body_typed([](CompileEngine self) { return self->LowerExternalFunctions(); });

TVM_REGISTER_GLOBAL("relay.backend._CompileEngineJIT")
    .set_body_typed([](CompileEngine self, CCacheKey key) { return self->JIT(key); });

TVM_REGISTER_GLOBAL("relay.backend._CompileEngineListItems").set_body_typed([](CompileEngine self) {
  CompileEngineImpl* ptr = dynamic_cast<CompileEngineImpl*>(self.operator->());
  ICHECK(ptr != nullptr);
  return ptr->ListItems();
});

TVM_REGISTER_GLOBAL("relay.backend._CompileEngineListShapeFuncItems")
    .set_body_typed([](CompileEngine self) {
      CompileEngineImpl* ptr = dynamic_cast<CompileEngineImpl*>(self.operator->());
      ICHECK(ptr != nullptr);
      return ptr->ListShapeFuncItems();
    });

TVM_REGISTER_GLOBAL("relay.backend._CompileEngineGetCurrentCCacheKey")
    .set_body_typed([](CompileEngine self) {
      CompileEngineImpl* ptr = dynamic_cast<CompileEngineImpl*>(self.operator->());
      ICHECK(ptr != nullptr);
      return ptr->GetCurrentCCacheKey();
    });

}  // namespace relay
}  // namespace tvm

File: python/tvm/relay/backend/compile_engine.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
# pylint: disable=len-as-condition,no-else-return
"""Backend code generation engine."""
from __future__ import absolute_import

import hashlib
import numpy as np
import tvm
from topi import tag
from ..base import register_relay_node, Object
from ... import _api_internal
from ... import target as _target
from ..._ffi.function import register_func
from ... import autotvm
from .. import expr as _expr
from .. import op as _op
from .. import ty as _ty
from ..expr_functor import ExprVisitor
from . import _backend

@register_relay_node
class CachedFunc(Object):
    """Low-level tensor function to back a relay primitive function.
    """
    def __init__(self, target, func_name, inputs, outputs, schedule=None,
                 lowered_funcs=None, shape_func_param_states=None):
        if lowered_funcs is None:
            lowered_funcs = []
        if shape_func_param_states is None:
            shape_func_param_states = []
        self.__init_handle_by_constructor__(
            _backend._make_CachedFunc, target, func_name, inputs, outputs,
            schedule, lowered_funcs, shape_func_param_states)


@register_relay_node
class CCacheKey(Object):
    """Key in the CompileEngine.

    Parameters
    ----------
    source_func : tvm.relay.Function
        The source function.

    target : tvm.Target
        The target we want to run the function on.
    """
    def __init__(self, source_func, target):
        self.__init_handle_by_constructor__(
            _backend._make_CCacheKey, source_func, target)


@register_relay_node
class CCacheValue(Object):
    """Value in the CompileEngine, including usage statistics.
    """


def _get_cache_key(source_func, target):
    if isinstance(source_func, _expr.Function):
        if isinstance(target, str):
            target = _target.create(target)
            if not target:
                raise ValueError("Need target when source_func is a Function")
        return CCacheKey(source_func, target)
    if not isinstance(source_func, CCacheKey):
        raise TypeError("Expect source_func to be CCacheKey")
    return source_func


def get_shape(shape):
    """Convert the shape to correct dtype and vars."""
    ret = []
    for dim in shape:
        if isinstance(dim, tvm.expr.IntImm):
            val = int(dim)
            assert val <= np.iinfo(np.int32).max
            ret.append(tvm.expr.IntImm("int32", val))
        elif isinstance(dim, tvm.expr.Any):
            ret.append(tvm.var("any_dim", "int32"))
        else:
            ret.append(dim)
    return ret


def get_valid_implements(op, attrs, inputs, out_type, target):
    """Get all valid implementations from the op strategy.

    Note that this function doesn't support op that has symbolic input shapes.

    Parameters
    ----------
    op : relay.op.Op
        Relay operator.

    attrs : object
        The op attribute.

    inputs : list of tvm.Tensor
        Input tensors to the op.

    out_type : relay.Type
        The output type.

    target : tvm.Target
        The target to compile the op.

    Returns
    -------
    ret : list of relay.op.OpImplement
        The list of op implementations.
    """
    fstrategy = op.get_attr("FTVMStrategy")
    assert fstrategy is not None, "%s doesn't have FTVMStrategy registered" % op.name
    with target:
        strategy = fstrategy(attrs, inputs, out_type, target)
    ret = []
    for spec in strategy.specializations:
        if spec.condition:
            # check if all the clauses in the specialized condition are true
            flag = True
            for clause in spec.condition.clauses:
                clause = tvm.ir_pass.Simplify(clause)
                if isinstance(clause, tvm.expr.IntImm) and clause.value:
                    continue
                flag = False
                break
            if flag:
                for impl in spec.implements:
                    ret.append(impl)
        else:
            for impl in spec.implements:
                ret.append(impl)
    return ret


def select_implement(op, attrs, inputs, out_type, target, use_autotvm=True):
    """Select the best implement from the op strategy.

    If use_autotvm is True, it'll first try to find the best implementation
    based on AutoTVM profile results. If no AutoTVM profile result is found,
    it'll choose the implementation with highest plevel.

    If use_autotvm is False, it'll directly choose the implementation with
    highest plevel.

    Note that this function doesn't support op that has symbolic input shapes.

    Parameters
    ----------
    op : relay.op.Op
        Relay operator.

    attrs : object
        The op attribute.

    inputs : list[tvm.Tensor]
        Input tensors to the op.

    out_type : relay.Type
        The output type.

    target : tvm.Target
        The target to compile the op.

    use_autotvm : bool
        Whether query AutoTVM to pick the best.

    Returns
    -------
    ret : tuple(relay.op.OpImplement, list[tvm.Tensor])
        The best op implementation and the corresponding output tensors.
    """
    all_impls = get_valid_implements(op, attrs, inputs, out_type, target)

    best_plevel_impl = None
    for impl in all_impls:
        if best_plevel_impl is None or int(impl.plevel) > int(best_plevel_impl.plevel):
            best_plevel_impl = impl
    if not use_autotvm:
        outs = best_plevel_impl.compute(attrs, inputs, out_type)
        return best_plevel_impl, outs

    outputs = {}
    best_autotvm_impl = None
    best_cfg = None
    dispatch_ctx = autotvm.task.DispatchContext.current
    for impl in all_impls:
        outs = impl.compute(attrs, inputs, out_type)
        outputs[impl] = outs
        workload = autotvm.task.get_workload(outs)
        if workload is None:
            continue
        cfg = dispatch_ctx.query(target, workload)
        if cfg.cost is None:
            # It's a fallback config
            continue
        if best_cfg is None or best_cfg.cost > cfg.cost:
            best_autotvm_impl = impl
            best_cfg = cfg
    if best_autotvm_impl:
        return best_autotvm_impl, outputs[best_autotvm_impl]
    return best_plevel_impl, outputs[best_plevel_impl]


class ScheduleGetter(ExprVisitor):
    """Get the schedule given a fused Relay function"""

    MAX_FUNC_NAME_LENGTH = 80

    def __init__(self, target):
        super().__init__()
        self.target = target
        self.master_op = None
        self.master_attrs = None
        self.master_op_pattern = 0
        self.master_implement = None
        self.func_name = ""
        self.scalars = []
        self._device_copy_op = _op.get("device_copy")

    def create(self, prim_func):
        """Get the schedule and create the cached function"""
        assert isinstance(prim_func, _expr.Function)
        assert prim_func.is_primitive()

        def create_tensors(typ, tensors):
            if isinstance(typ, _ty.TensorType):
                tensors.append(tvm.placeholder(get_shape(typ.shape), typ.dtype))
            else:
                assert isinstance(typ, _ty.TupleType)
                for field in typ.fields:
                    create_tensors(field, tensors)

        inputs = []
        for param in prim_func.params:
            tensors = []
            create_tensors(param.checked_type, tensors)
            self.memo_map[param] = tensors
            inputs.extend(tensors)
        self.func_name = "fused"
        outputs = self.visit(prim_func.body)
        if len(self.func_name) > ScheduleGetter.MAX_FUNC_NAME_LENGTH:
            hash_digest = int(hashlib.sha1(self.func_name).hexdigest(), 16)
            self.func_name = "%s_%s" % (
                self.func_name[:ScheduleGetter.MAX_FUNC_NAME_LENGTH], hash_digest)

        assert self.master_op is not None
        tensor_outs = []
        for tensor in outputs:
            if not isinstance(tensor.op, tvm.tensor.PlaceholderOp):
                tensor_outs.append(tensor)
        sch = None
        if not isinstance(self.master_attrs, _op.op_attrs.DeviceCopyAttrs):
            # print('master op:', self.master_op.name)
            sch = self.master_implement.schedule(self.master_attrs, tensor_outs, self.target)
            for scalar in self.scalars:
                sch[scalar].compute_inline()
        return CachedFunc(self.target, self.func_name, inputs, outputs, sch)

    def visit_var(self, var):
        assert False, "Found free variable " + var.name_hint

    def visit_constant(self, const):
        assert len(const.data.shape) == 0, "Constant is not scalar"
        dtype = const.data.dtype
        data = const.data.asnumpy()
        def fcompute():
            if dtype.startswith("int"):
                return tvm.expr.IntImm(dtype, int(data))
            elif dtype.startswith("uint"):
                return tvm.expr.UIntImm(dtype, int(data))
            elif dtype.startswith("float"):
                return tvm.expr.FloatImm(dtype, float(data))
            else:
                assert False, "not handled"
                return tvm.expr.Expr()
        value = tvm.compute((), fcompute, name="compile_engine_const", tag=tag.BROADCAST)
        self.scalars.append(value.op)
        return [value]

    def visit_call(self, call):
        inputs = []
        count_tuple = 0
        for arg in call.args:
            if isinstance(arg.checked_type, _ty.TupleType):
                count_tuple += 1
            inputs.extend(self.visit(arg))
        assert count_tuple <= 1, "Only allow function with a single tuple input"
        ret_type = call.checked_type
        if isinstance(ret_type, _ty.TensorType):
            ret_type = _ty.TensorType(get_shape(ret_type.shape), ret_type.dtype)
        elif isinstance(ret_type, _ty.TupleType):
            new_fields = []
            for field in ret_type.fields:
                if isinstance(field, _ty.TensorType):
                    new_fields.append(_ty.TensorType(get_shape(field.shape), field.dtype))
                else:
                    new_fields.append(field)
            ret_type = _ty.TupleType(new_fields)
        assert isinstance(call.op, _op.Op)
        op = call.op

        # disable AutoTVM tracing if op is not in wanted list
        env = autotvm.task.TaskExtractEnv.current
        reenable_tracing = False
        if env is not None and env.tracing:
            if env.wanted_relay_ops is not None and op not in env.wanted_relay_ops:
                env.tracing = False
                reenable_tracing = True

        if op == self._device_copy_op:
            copy_input = inputs[0]
            outputs = [_api_internal._Tensor(copy_input.shape, copy_input.dtype,
                                             None, 0)]
        else:
            is_dyn = call.checked_type.is_dynamic()
            for arg in call.args:
                is_dyn = is_dyn or arg.checked_type.is_dynamic()

            if not is_dyn:
                best_impl, outputs = select_implement(
                    op, call.attrs, inputs, ret_type, self.target)
            else:
                # TODO(@icemelon9): Allow tvm to generate multiple kernels for dynamic shapes
                # for dynamic case, we currently use the implementation with highest plevel
                best_impl, outputs = select_implement(
                    op, call.attrs, inputs, ret_type, self.target, use_autotvm=False)
        op_pattern = op.get_attr("TOpPattern")
        if op_pattern >= _op.OpPattern.COMM_REDUCE:
            assert self.master_op is None or self.master_op_pattern < _op.OpPattern.COMM_REDUCE, \
                "Two complicated op in a primitive function master=%s current=%s" % (
                    self.master_op, op)
        if op_pattern >= self.master_op_pattern:
            self.master_op = op
            self.master_attrs = call.attrs
            self.master_op_pattern = op_pattern
            self.master_implement = best_impl
        if len(outputs) > 1:
            assert isinstance(call.checked_type, _ty.TupleType)
            assert len(call.checked_type.fields) == len(outputs)
        if op == self._device_copy_op:
            self.func_name += "__copy"
        else:
            self.func_name += "_" + op.name

        # re-enable AutoTVM tracing
        if reenable_tracing:
            env.tracing = True

        return outputs

    def visit_let(self, let):
        val = self.visit(let.value)
        assert let.var not in self.memo_map
        self.memo_map[let.var] = val
        return self.visit(let.body)

    def visit_tuple(self, tup):
        fields = []
        for field in tup.fields:
            assert isinstance(field.checked_type, _ty.TensorType), "Only allow Tuple of Tensor"
            res = self.visit(field)
            assert len(res) == 1
            fields.append(res[0])
        return fields

    def visit_tuple_getitem(self, t):
        tup = self.visit(t.tuple)
        assert len(tup) == len(t.tuple.checked_type.fields)
        assert t.index >= 0
        assert t.index < tup.size()
        return [tup[t.index]]


@register_func("relay.backend.create_schedule")
def create_schedule(src_func, target):
    return ScheduleGetter(target).create(src_func)


@register_relay_node
class CompileEngine(Object):
    """CompileEngine to get lowered code.
    """
    def __init__(self):
        raise RuntimeError("Cannot construct a CompileEngine")

    def lower(self, source_func, target=None):
        """Lower a source_func to a CachedFunc.

        Parameters
        ----------
        source_func : Union[tvm.relay.Function, CCacheKey]
            The source relay function.

        target : tvm.Target
            The target platform.

        Returns
        -------
        cached_func: CachedFunc
            The result of lowering.
        """
        # pylint: disable=broad-except, import-outside-toplevel
        try:
            key = _get_cache_key(source_func, target)
            return _backend._CompileEngineLower(self, key)
        except Exception:
            import traceback
            msg = traceback.format_exc()
            msg += "Error during compile func\n"
            msg += "--------------------------\n"
            msg += source_func.astext(show_meta_data=False)
            msg += "--------------------------\n"
            raise RuntimeError(msg)

    def lower_shape_func(self, source_func, target=None):
        key = _get_cache_key(source_func, target)
        return _backend._CompileEngineLowerShapeFunc(self, key)

    def jit(self, source_func, target=None):
        """JIT a source_func to a tvm.runtime.PackedFunc.

        Parameters
        ----------
        source_func : Union[tvm.relay.Function, CCacheKey]
            The source relay function.

        target : tvm.Target
            The target platform.

        Returns
        -------
        jited_func: tvm.runtime.PackedFunc
            The result of jited function.
        """
        key = _get_cache_key(source_func, target)
        return _backend._CompileEngineJIT(self, key)

    def clear(self):
        """clear the existing cached functions"""
        _backend._CompileEngineClear(self)

    def items(self):
        """List items in the cache.

        Returns
        -------
        item_list : List[Tuple[CCacheKey, CCacheValue]]
            The list of items.
        """
        res = _backend._CompileEngineListItems(self)
        assert len(res) % 2 == 0
        return [(res[2*i], res[2*i+1]) for i in range(len(res) // 2)]

    def dump(self):
        """Return a string representation of engine dump.

        Returns
        -------
        dump : str
            The dumped string representation
        """
        items = self.items()
        res = "====================================\n"
        res += "CompilerEngine dump, %d items cached\n" % len(items)
        for k, v in items:
            res += "------------------------------------\n"
            res += "target={}\n".format(k.target)
            res += "use_count={}\n".format(v.use_count)
            res += "func_name={}\n".format(v.cached_func.func_name)
            res += k.source_func.astext() + "\n"
        res += "===================================\n"
        return res


def get():
    """Get the global compile engine.

    Returns
    -------
    engine : tvm.relay.backend.CompileEngine
        The compile engine.
    """
    return _backend._CompileEngineGlobal()
